frontend/src/utils/generate-warnings.ts

```diff
@@ -75,6 +90,473 @@ export function produceWarnings(schedule: Schedule): IWarning[] {
+  }
+
+  return Array.from(reqGroups);
+}
+
+/**
+ * Identify unsatisfied requirements given a major and a schedule.
+ * @param schedule the schedule to check requirements for.
+ * @param major the major to check requirements against.
+ */
+export function produceRequirementGroupWarning(
+  schedule: Schedule,
+  major: Major
+): IRequirementGroupWarning[] {
+  // holds courses that are currently on the schedule.
+  const taken: Map<string, number> = new Map<string, number>();
+  const coursesUsed: Set<string> = new Set<string>();
+
+  //add courses from the schedule to a Map: string => number (produceCourseCode => creditHours)
+  schedule.years.sort((n1, n2) => n1 - n2);
+  for (const yearNum of schedule.years) {
+    const year = schedule.yearMap[yearNum];
+
+    // add courses, term by term.
+    addCoursesToMap(year.fall, taken);
+
+    addCoursesToMap(year.spring, taken);
+
+    // if is summer full, only check summer I.
+    // shouldn't matter, summerII would just be empty.
+    if (year.isSummerFull) {
+      addCoursesToMap(year.summer1, taken);
+    } else {
+      addCoursesToMap(year.summer1, taken);
+      addCoursesToMap(year.summer2, taken);
+    }
+  }
+
+  let requirementGroups: string[] = major.requirementGroups;
+  let res: IRequirementGroupWarning[] = [];
+  for (const name of requirementGroups) {
+    let requirementGroup: IMajorRequirementGroup =
+      major.requirementGroupMap[name];
+    // todo: if req group is RangeSection or contains a ICourseRange process after all other requirement groups.
+    // this is because these reqs can be satisfied by a wide range of courses, and you don't want it using up
+    // a course that is need to satisfy a more constrained requirement.
+    let unsatisfiedRequirement:
+      | IRequirementGroupWarning
+      | undefined = produceUnsatifiedRequirement(
+      requirementGroup,
+      taken,
+      coursesUsed
+    );
+    if (unsatisfiedRequirement) {
+      res.push(unsatisfiedRequirement);
+    }
+  }
+  return res;
+}
+
+/**
+ * Produce an IUnsatisfiedRequirementGroup object if the requirementGroup hasn't been fully satisfied. Undefined otherwise.
+ * @param requirementGroup the requirement group to check.
+ * @param taken the Map of courses a student has on their schedule right now.
+ * @param coursesUsed a set of courses which already satisfy some requirement for the major.
+ */
+function produceUnsatifiedRequirement(
+  requirementGroup: IMajorRequirementGroup,
+  taken: Map<string, number>,
+  coursesUsed: Set<string>
+): IRequirementGroupWarning | undefined {
+  switch (requirementGroup.type) {
+    case "AND": {
+      return processAndSection(requirementGroup, taken, coursesUsed);
+    }
+    case "OR": {
+      return processOrSection(requirementGroup, taken, coursesUsed);
+    }
+    case "RANGE": {
+      return processRangeSection(requirementGroup, taken, coursesUsed);
+    }
+    default: {
+      return undefined;
```

Comment:
Thought about this a little more, and decided I think the move is to throw here.

Reply:
I think it is okay return undefined here because we do not `"produceUnsatifiedRequirement"` for a requirement group that we do not recognize. But I can throw too.